### PR TITLE
The BBCode conversion is split into several smaller functions

### PR DESCRIPTION
**Pull request by annando**
This PR shouldn't have got any influence on the conversion result. But it improves the readabilty of the `convert` function. It also adds some new BBCodes, based on their HTML counterparts. They aren't added to the documentation at this moment, so that people aren't using them right now. They are now used in the HTML2BBCode functionality to "improve the visual experience".

### CODE REVIEW

File: tests/src/Factory/Api/Twitter/DirectMessageTest.php [hunk 1]
```diff
@@ -67,7 +67,7 @@ public function testApiFormatMessagesWithHtmlText()
 			->toArray();
 
 		self::assertEquals('item_title', $directMessage['title']);
-		self::assertEquals('<strong>item_body</strong>', $directMessage['text']);
+		self::assertEquals('<b>item_body</b>', $directMessage['text']);
```
Comment (MrPetovan) [inline tests/src/Factory/Api/Twitter/DirectMessageTest.php:70]: Why the change?

File: src/Content/Text/BBCode.php [hunk 2]
```diff
@@ -1341,581 +1339,611 @@ public static function convert(string $text = null, bool $try_oembed = true, int
...
-				$SpoilerLayout = '<details class="spoiler"><summary>' . DI::l10n()->t('Click to open/close') . '</summary>$1</details>';
+	private static function convertStylesToHtml(string $text, int $simple_html): string
+	{
+		// Several easy to replace HTML elements
+		// @todo add the new elements to the documentation by the end of 2024 so that most systems will support them.
+		$elements = ['b', 'del', 'em', 'i', 'ins', 'kbd', 'mark',
+			's', 'samp', 'small', 'strong', 'sub', 'sup', 'u', 'var'];
+		foreach ($elements as $element) {
+			$text = preg_replace("(\[" . $element . "\](.*?)\[\/" . $element . "\])ism", '<' . $element . '>$1</' . $element . '>', $text);
+		}
```
Comment (MrPetovan) [inline src/Content/Text/BBCode.php:1668]: I don't think the `[i]` and `[em]` tags should convert to different HTML tags because we don't differentiate their use in the BBCode documentation. As a result, I believe we should be able to specify an alternative HTML tag a BBCode should convert to.
```suggestion
		$elements = ['b', 'del' => 's', 'em' => 'i', 'i', 'ins', 'kbd', 'mark',
			's', 'samp', 'small', 'strong' => 'b', 'sub', 'sup', 'u', 'var'];
		foreach ($elements as $source => $element) {
			$source = is_string($source) ? $source : $element;
			$text = preg_replace("(\[" . $source . "\](.*?)\[\/" . $source . "\])ism", '<' . $element . '>$1</' . $element . '>', $text);
		}
```

**Review by MrPetovan — CHANGES_REQUESTED** [review]

File: tests/src/Factory/Api/Twitter/DirectMessageTest.php (hunk 1 above)
Comment (annando) [reply tests/src/Factory/Api/Twitter/DirectMessageTest.php:70]: There is now a dedicated "strong" element, so there is "b" *and* "strong".

File: src/Content/Text/BBCode.php (hunk 2 above)
Comment (annando) [reply src/Content/Text/BBCode.php:1668]: "i" and "em" are different tags, although they look the same. The same goes with "del" and "s". Especially for this element we could use some changed CSS for this element.

File: tests/src/Factory/Api/Twitter/DirectMessageTest.php (hunk 1 above)
Comment (MrPetovan) [reply tests/src/Factory/Api/Twitter/DirectMessageTest.php:70]: Why do we need both? Do we discriminate between both in our display? Do we expect remote platforms to discriminate between both? I'd rather we keep it simpler.

File: src/Content/Text/BBCode.php (hunk 2 above)
Comment (MrPetovan) [reply src/Content/Text/BBCode.php:1668]: We could but we don't at the moment, and we're presenting formatting options rather than semantic elements in the post compose form, so I believe we should stick to the corresponding HTML tags `<i>` and `<s>`.

File: tests/src/Factory/Api/Twitter/DirectMessageTest.php (hunk 1 above)
Comment (annando) [reply tests/src/Factory/Api/Twitter/DirectMessageTest.php:70]: `b` and `strong` are different elements and neither of them is deprecated, see here: https://developer.mozilla.org/en-US/docs/Web/HTML/Element

They could be treated differently by for example screenreaders. It's the same like just using bigger font sizes instead of using some header elements like `h1`, `h2`, ... Optically it looks the same, but technically it is different.

File: src/Content/Text/BBCode.php (hunk 2 above)
Comment (annando) [reply src/Content/Text/BBCode.php:1668]: I would like to use some CSS like the one used here https://developer.mozilla.org/en-US/docs/Web/HTML/Element/del or here: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/ins in the future.

File: tests/src/Factory/Api/Twitter/DirectMessageTest.php (hunk 1 above)
Comment (MrPetovan) [reply tests/src/Factory/Api/Twitter/DirectMessageTest.php:70]: I didn't say any of them was deprecated, I'm saying `<strong>` isn't relevant for us because our post body editor isn't semantic-centric, it's formatting-centric. So we should stick to `<b>`.

File: src/Content/Text/BBCode.php (hunk 2 above)
Comment (MrPetovan) [reply src/Content/Text/BBCode.php:1668]: That's cool, but we can't assume remote systems will have adequate support for any of them, so I believe we should stick to the lowest common denominator tags, `<i>` and `<s>`. I guess that's the logical downside of trying to be compatible with as many different protocols as possible.

Comment (annando) [reply src/Content/Text/BBCode.php:1668]: We can import web pages via the feed functionality. BTW: I tested all these elements with Mastodon and Pleroma. Pleroma for example doesn't render `s`, but `del`.

Friendica:
image

Pleroma:
image

We can't assume, that remote systems will render elements that look identical on Friendica will look identical on their system. So we shouldn't assume that.

File: tests/src/Factory/Api/Twitter/DirectMessageTest.php (hunk 1 above)
Comment (annando) [reply tests/src/Factory/Api/Twitter/DirectMessageTest.php:70]: We can define headlines, which definitely is semantic and not formatting. But anyway: I don't see a reason not to expand the capabilities here.

**Review by MrPetovan — APPROVED** [review]
